### ISSUE TITLE
Blazor wasm dynamic linking native sidemodules

### ISSUE THREAD

**Issue by Hazardu** [dotnet/runtime#75257]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

I am trying to link a C++ library compiled with Emscripten to a Blazor wasm application, and use pinvoke. Thus far, the only way to link the two is to use static linking via `NativeFileReference`, but emcc contains functionality to dynamically link side modules and a main module.

Comparing the generated JS in an app where both the lib and main app are written in C++ to one with Blazor Wasm and `NativeFileReference`, it seems that the code responsible for loading the functions into the `Module` object is missing, thus when the app tries to call them when they are uninitialized and `abort` is called and the whole thing breaks.

Trying a different approach of statically linking a part of the C++ code containing C wrapper functions that are PInvoked and dynamically linking the C++ class that is wrapped, the same issue happened after the C function reached a function call. 

### Describe the solution you'd like

`NativeFileReference` other than accepting relocatable wasm files for static linking, would accept native libs compiled with Emscripten's SIDE_MODULE flag and expand the generated javascript to load them at loadtime.

### Additional context

https://github.com/Devsh-Graphics-Programming/GPU-With-C-Sharp-Angular-WASM
Thus far this app is on the small end, but we are considering making a robust native wasm library that contains the Nabla engine. For that dynamic linking is preferable.

Emscripten doc
https://emscripten.org/docs/compiling/Dynamic-Linking.html

**Comment by radical** [comment]
cc @radekdoulik @pavelsavara @SteveSandersonMS @kg

**Comment by BenMcLean** [comment]
It looks like this is the primary issue blocking Godot **_and_** Stride **_and_** MonoGame **_and_** probably most/any other FOSS game engines and frameworks that use C# from building programs for WebAssembly on .NET Core 5+. It would be nice for it to get more attention.

**Comment by GeorgeS2019** [comment]
@BenMcLean 
>  this is the **primary issue blocking**  Godot 4 from building programs for WebAssembly on .NET Core 5+.

Will [godot-dotnet](https://github.com/raulsntos/godot-dotnet) [helps ](https://github.com/godotengine/godot/issues/70796#issuecomment-2121119041) to resolve this?

**Comment by GeorgeS2019** [comment]
@BenMcLean 
https://chat.godotengine.org/channel/dotnet?msg=ow94ZyspFmyvvzGYh

> There's also ongoing work on the .NET runtime to use a new **NativeAOT-LLVM** runtime for the web platform which may solve our problem, but it's very early and very experimental. Not sure how much of this will make it to .NET 9.0.

Anyone can share links to **NativeAOT-LLVM**?

**Comment by Delsin-Yu** [comment]
@GeorgeS2019 
> Will [godot-dotnet](https://github.com/raulsntos/godot-dotnet) [helps ](https://github.com/godotengine/godot/issues/70796#issuecomment-2121119041) to resolve this?

No, It won't.

**Comment by BenMcLean** [comment]
@lewing Why did this get pushed into the indefinite future?

**Comment by lewing** [comment]
The limitations of emscripten's dynamic linking have made it quite challenging for us to use in practice.  Can you explain in more detail what problem you are attempting to solve?  We use feedback to help prioritize work.

As for browser-wasm threading it is functional and you can try in out in recent net9.0 previews, but only the runtime side and that will remain experimental in net9. The Blazor front end work as indicated in https://github.com/dotnet/aspnetcore/issues/17730 is part of the planning for net10.

cc @danroth27

**Comment by BenMcLean** [comment]
This comment here offered an explanation of how lack of dynamic linking native sidemodules prevents Godot 4 from supporting WASM export: https://github.com/godotengine/godot/issues/70796#issuecomment-1618006609

I believe Stride has a similar issue.

**Comment by JeroMiya** [comment]
I'm not an expert on the internals of Stride, but I don't believe this issue affects Stride, at least not in the same way as Godot. Web export doesn't exist yet in Stride, but if it were implemented similarly to the Desktop projects, the C# assembly is the main entry point into the application. Stride itself is a NuGet package you reference in your platform specific project. So theoretically, a WASM export should be possible - I think the missing piece is stable WebGL/WebGPU bindings for WASM/.NET, and of course developer resources to actually implement the web export functionality.

I think the Godot-as-library route is probably the best route for that engine, not just for WASM but also for more robust iOS/Android support and likely console support as well.

**Comment by Delsin-Yu** [comment]
@lewing 
> The limitations of emscripten's dynamic linking have made it quite challenging for us to use in practice. Can you explain in more detail what problem you are attempting to solve? We use feedback to help prioritize work.

The Core issue for the Godot Game Engine is that we are embedding the dotnet into a C++ engine environment (either by hostfxr or loading NAOT pre-compiled library), which means there are bound to be two separate modules, the Godot Wasm and the dotnet Wasm+user assembly(s); and, as [the comment here](https://github.com/godotengine/godot/issues/70796#issuecomment-2524834258) explains, currently there is no way to dynamically link them (neither dotnet wasm as the main module nor dotnet wasm as the side module) like what we can do in every other platform.

**Comment by Delsin-Yu** [comment]
Hello, does the .NET Wasm team still remember this issue? Is there any shareable information or development I can refer to?

**Comment by vonweller** [comment]
Has an effective solution already been found?

**Comment by Mohsen7s** [comment]
"Future" milestone made the issue to looks like a ignorat issue. Already hundreds of godot game developers are blocked by this issue. Please re-add a specified milestone.

**Comment by AlanDavison** [comment]
Yeah, I don't want to apply too much forceful pressure, because I _get_ how priorities are set.

Perhaps it would be helpful to have all of the major blocking problems collated somewhere so that capable people can at least begin working on a prototype version. Maybe that would make the decision to move this higher up the priority list easier?

**Comment by hez2010** [comment]
IIRC this is already possible with NativeAOT-LLVM (an experimental branch that enables NativeAOT on wasm), where you can export functions in NativeAOT-compiled wasm natively. Now that mono is going to be retired and the .NET team is bringing up early CoreCLR/R2R support to wasm in .NET 11, I think it might be a good time to finally push this thing forward. 

(.NET 11 is already near code-complete so at least we will have to wait for .NET 12)
